### PR TITLE
FEATURE(serve): add --failed-only to sky serve down to bulk-remove failed replicas (#7319)

### LINKED ISSUE #7319: FAILED_PROVISION is bloating the sky serve status output
My pod was having issues resolving / provisioning and now I have hundreds of FAILED_PROVISION status service replicas that make it difficult to monitor the successful / ready replicas after I fixed the issue. Is it possible to clear these FAILED_PROVISION status replicas?

### PR DESCRIPTION
**Pull request by pravarpathania**
# Problem
Users accumulate many `FAILED_PROVISION` (and other failed) replicas that clutter `sky serve status` and are impractical to remove one-by-one.

# Changes

## Refactoring (Addressing Review Feedback)
Following @cblmemo and @andylizf's feedback, this implementation has been refactored to:
- Rename `--failed-only` to `--failed-replicas` for clearer semantics
- Consolidate into existing `terminate_replica()` API instead of creating a separate function
- Use sentinel value (`replica_id=-1`) to indicate "all failed replicas" (avoids breaking API changes)
- Fix `_async_call_or_wait()` request naming to use `'serve.terminate_replica'`

**Client-Server Flow:**
1. CLI validates arguments → 2. SDK makes authenticated request → 3. Server SSHs to controller → 4. Code executes on controller (database access) → 5. Results returned to user

## Files Modified
- **`sky/client/cli/command.py`**: Added `--failed-replicas` flag with validation (single service, no `--all`/`--replica-id`), confirmation prompt, calls unified `terminate_replica()` API
- **`sky/serve/serve_utils.py`**: Extended `terminate_replica()` to handle both single and bulk termination via `failed_replicas` parameter
- **`sky/serve/server/core.py`**: Extended `terminate_replica()` to handle failed replicas bulk termination
- **`sky/serve/client/sdk.py`**: Extended `terminate_replica()` signature with `failed_replicas` parameter
- **`sky/serve/server/server.py`**: Updated `/terminate-replica` endpoint to handle both cases
- **`sky/server/requests/payloads.py`**: Extended `ServeTerminateReplicaBody` with `failed_replicas` field
- **`tests/test_serve_down_failed_only.py`**: 16 tests covering CLI, server-side, integration, and edge cases
- **`tests/smoke_tests/test_sky_serve.py`**: Added 2 end-to-end smoke tests

# Usage
```bash
sky serve down <service_name> --failed-replicas [-y]
```
**Behavior:**

* Terminates all replicas in failed states (`FAILED_PROVISION`, `FAILED`, `FAILED_CLEANUP`, `FAILED_INITIAL_DELAY`, `FAILED_PROBING`, `UNKNOWN`)  
* Uses `purge=True` for forceful cleanup  
* Continues on per-replica errors  
* Returns detailed summary

# **Testing**

**Unit Tests (16/16 passing locally):**

* CLI flag validation (cannot use with `--all` or `--replica-id`)  
* Confirmation prompt behavior  
* API parameter validation  
* Server-side logic with mocked database  
* Integration tests for complete flow  
* Edge cases (no replicas, service not found)

**Smoke Tests (added, need CI):**

* `test_skyserve_down_failed_replicas` \- Tests cleanup of failed replicas using existing `user_bug.yaml`  
* `test_skyserve_down_failed_replicas_noop` \- Tests no-op behavior with healthy service using existing `auto_restart.yaml`

I don't have cloud access to run smoke tests locally. Could you help trigger them via `/smoke-test -k failed_replicas`?

- Fixes: #7319 

- [ ] Code formatting: install pre-commit (auto-check on commit) or `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `/smoke-test` (CI) or `pytest tests/test_smoke.py` (local)
- [ ] Relevant individual tests: `/smoke-test -k test_name` (CI) or `pytest tests/test_smoke.py::test_name` (local)
- [ ] Backward compatibility: `/quicktest-core` (CI) or `pytest tests/smoke_tests/test_backward_compat.py` (local)

### CODE REVIEW

File: sky/client/cli/command.py [hunk 1]
```diff
@@ -5697,6 +5697,78 @@ def serve_status(verbose: bool, endpoint: bool, service_names: List[str]):
...
+    from sky.serve import serve_state
+
+    # Get all failed replica statuses
+    failed_statuses = serve_state.ReplicaStatus.failed_statuses()
+
+    # Collect all failed replicas, using a dict to deduplicate by replica_id
+    failed_replicas_dict = {}
+    try:
+        for replica_status in failed_statuses:
+            replicas = serve_state.get_replicas_at_status(
```
Comment (andylizf) [inline sky/client/cli/command.py]: iiuc this logic should be run on serve controller, not the local.

Comment (andylizf) [reply sky/client/cli/command.py]: can you take a look at the sky.serve.status function?

**Review by andylizf — CHANGES_REQUESTED** [review]

File: sky/client/cli/command.py (hunk 1 above)
Comment (pravarpathania) [reply sky/client/cli/command.py]: Thanks for pointing this out @andylizf. I’ll look into this and update the PR so the logic runs on the controller instead of locally.

**Comment by pravarpathania** [comment]
I have updated code in the PR, kindly let me know if it is fine or any other changes are required.

File: sky/client/cli/command.py [hunk 2]
```diff
@@ -5792,15 +5823,21 @@ def serve_down(
                           abort=True,
                           show_default=True)
 
-    if replica_id_is_defined:
+    if failed_only:
+        # Handle --failed-only: terminate all failed replicas
+        service_name = service_names[0]
+        request_id = serve_lib.terminate_failed_replicas(service_name)
+        _async_call_or_wait(request_id, async_call, 'sky.serve.down')
```
Comment (andylizf) [inline sky/client/cli/command.py]: should be serve_lib.terminate_failed_replicas?

**Review by andylizf — CHANGES_REQUESTED** [review]
Mostly LGTM. Also, could we have some smoke tests? IIUC they are all simple unit tests now, and don't cover the core logic.

File: sky/client/cli/command.py (hunk 2 above)
Comment (pravarpathania) [reply sky/client/cli/command.py]: @andylizf Thank you for the review!

---
Confirmed — line 5828 correctly calls:

```python
request_id = serve_lib.terminate_failed_replicas(service_name)
```
This function is defined in sky/serve/client/sdk.py (lines 157–183) and follows the same pattern as other serve SDK functions like terminate_replica() and down():

```python
@usage_lib.entrypoint
@server_common.check_server_healthy_or_start
def terminate_failed_replicas(service_name: str) -> server_common.RequestId[None]:
    """Terminates all failed replicas for the given service."""
    body = payloads.ServeTerminateFailedReplicasBody(service_name=service_name)
    response = server_common.make_authenticated_request(
        'POST', '/serve/terminate-failed-replicas',
        json=json.loads(body.model_dump_json()),
        timeout=(5, None))
    return server_common.get_request_id(response)
```
It’s properly exported in sky/serve/__init__.py and follows the standard
SDK → request_id → async_wait pattern used throughout the codebase.
If you see any issues with this approach, please let me know what I should fix.

---
### **Re: Smoke Tests**
I’ll create end-to-end smoke tests that:

Deploy actual services with replicas

Force replicas into failed states (FAILED_PROVISION, FAILED_PROBING, etc.)

Verify only failed replicas are terminated while healthy ones remain

Comment (andylizf) [reply sky/client/cli/command.py]: well i mean the argument in _async_call_or_wait

File: sky/client/cli/command.py [hunk 3]
```diff
@@ -5711,6 +5711,11 @@ def serve_status(verbose: bool, endpoint: bool, service_names: List[str]):
               default=None,
               type=int,
               help='Tear down a given replica')
+@click.option('--failed-only',
```
Comment (cblmemo) [inline sky/client/cli/command.py]: I feel like this should be renamed to sth like `--failed-replicas` because the default operation of this command line is to terminate a service.

File: sky/serve/serve_utils.py [hunk 4]
```diff
@@ -1634,6 +1733,14 @@ def terminate_replica(cls, service_name: str, replica_id: int,
         ]
         return cls._build(code)
 
+    @classmethod
+    def terminate_failed_replicas(cls, service_name: str) -> str:
```
Comment (cblmemo) [inline sky/serve/serve_utils.py]: Can we repurpose the `terminate_replica` function instead of introducing a brand new api call for this?

File: sky/client/cli/command.py (hunk 2 above)
Comment (pravarpathania) [reply sky/client/cli/command.py]: You're right, thanks for catching that! The third argument should be 'serve.terminate_failed_replicas' to match what the operation is actually doing. I'm actually going to refactor this based on @cblmemo's suggestion to repurpose the existing terminate_replica function instead of adding a new API, so this will be addressed in that update.

File: sky/serve/serve_utils.py (hunk 4 above)
Comment (pravarpathania) [reply sky/serve/serve_utils.py]: Thanks for the feedback! Both suggestions make sense. Renaming to --failed-replicas is much clearer, and repurposing terminate_replica() instead of creating a separate API is a better approach. It reduces duplication and keeps the codebase cleaner.
I'll refactor the code to extend the existing terminate_replica() function to handle both cases. Will update the PR soon with these changes.

**Comment by pravarpathania** [comment]
## Refactoring Complete

I've refactored the implementation based on review feedback:

**Key Changes:**
1. Renamed `--failed-only` to `--failed-replicas`
2. Consolidated into existing `terminate_replica()` API with `failed_replicas` parameter
3. Used sentinel value (`replica_id=-1`) to maintain backwards compatibility (no API version bump needed)
4. Fixed `_async_call_or_wait()` request naming

**Testing:**
- Unit tests: 16/16 passing locally
- Smoke tests: Added 2 tests but need CI to run them

Please let me know if the approach looks good or if any further changes are needed. 
Once approved, could you trigger the smoke tests with `/smoke-test -k failed_replicas`? I don't have cloud access to run them locally.

**Comment by pravarpathania** [comment]
Hi guys @cblmemo @andylizf, can you please check the code let me know if it is fine now or I need to make further changes.
